### PR TITLE
Added emacs as an "operating system" for input mode. 

### PR DESCRIPTION
**Pull request by chewxy**
I added a new unicode input mode: emacs. 

While strictly not an operating system, inserting unicode characters in emacs has its own keybinding. The most usual way is `C-x-8 RET` followed by the unicode in hex. As a person who works across different OSes but with the constant being emacs, this was the most straightforward way forwards. 

Of course, the main credit goes to TrentinQuarantino in this [emacs stack exchange post](https://emacs.stackexchange.com/a/56027).
 

## Description

I added a new unicode input mode in `quantum/process_unicode_common.(c|h)`. Then I updated all the programmer friendly numbers of the enums, and added documentation too.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

## Additional notes

I'm unfamiliar with the testing protocols of this project. If I could get some help that'd be great. I'm reasonably familiar with C. However, what I have done is I have flashed two ergodox-EZs with this and tested it on emacs on both linux and macos. My Windows computer isn't quite available right now, so I'm unable to test on Windows.

### CODE REVIEW

File: quantum/quantum_keycodes.h [hunk 1]
```diff
@@ -385,113 +385,114 @@ enum quantum_keycodes {
     UNICODE_MODE_WIN,     // 5CE4
     UNICODE_MODE_BSD,     // 5CE5
     UNICODE_MODE_WINC,    // 5CE6
+    UNICODE_MODE_EMACS,   // SCE7
```
Comment (zvecr) [inline quantum/quantum_keycodes.h:388]: Keycodes cannot be added to the middle of the list as it breaks existing assumptions in things like via

**Review by zvecr — CHANGES_REQUESTED** [review]

**Comment by zvecr** [comment]
Previous implementation, #8358.

**Comment by drashna** [comment]
Is there a reason that the "linux" mode can't be used, with a custom start/finish option?

Eg

```c
void unicode_input_start(void) {
    unicode_saved_caps_lock = host_keyboard_led_state().caps_lock;
    unicode_saved_num_lock  = host_keyboard_led_state().num_lock;

    if (unicode_config.input_mode == UC_LNX && unicode_saved_caps_lock) {
        tap_code(KC_CAPS_LOCK);
    }

    unicode_saved_mods = get_mods(); // Save current mods
    clear_mods();                    // Unregister mods to start from a clean state

    switch (unicode_config.input_mode) {
        case UC_MAC:
            register_code(UNICODE_KEY_MAC);
            break;
        case UC_LNX:
            tap_code16(LCTL(KC_X));
            tap_code16(KC_8);
            tap_code16(KC_ENTER);
            break;
        case UC_WIN:
            // For increased reliability, use numpad keys for inputting digits
            if (!unicode_saved_num_lock) {
                tap_code(KC_NUM_LOCK);
            }
            register_code(KC_LEFT_ALT);
            wait_ms(UNICODE_TYPE_DELAY);
            tap_code(KC_KP_PLUS);
            break;
        case UC_WINC:
            tap_code(UNICODE_KEY_WINC);
            tap_code(KC_U);
            break;
    }

    wait_ms(UNICODE_TYPE_DELAY);
}
```

**Comment by chewxy** [comment]
@drashna as far as I am aware, ibus on linux only allows Alt+HEXCODE entries, whereas emacs in general requires the `C-x-8 RET` sequence followed by the hex code

**Comment by chewxy** [comment]
(having said that, if you have `gnome-tweak-tool` or `gconf-editor` on any linux with gnome (e.g. ubuntu), you can turn on Emacs Input and the Ctrl+x+8 ENTER sequence also allows you to type unicode.)

But we'd probably still want to keep the default ibus settings for normal people

File: quantum/quantum_keycodes.h (hunk 1 above)
Comment (chewxy) [reply quantum/quantum_keycodes.h:388]: OK I can change that

Comment (zvecr) [reply quantum/quantum_keycodes.h:388]: Now exists twice, along with all the comment changes.

Comment (chewxy) [reply quantum/quantum_keycodes.h:388]: Drat. Careless fix

Comment (chewxy) [reply quantum/quantum_keycodes.h]: This has been fixed now. I had ran `git checkout`  from the wrong source.

**Review by tzarc — APPROVED** [review]

**Review by zvecr — APPROVED** [review]

**Pull request merged**